plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		return nil
+	}
+
+	var telemetryName string
+	if useFieldNameInTelemetryName {
+		telemetryName = metric.Name() + "_" + fieldName
+	} else {
+		telemetryName = metric.Name()
+	}
+	telemetry := appinsights.NewMetricTelemetry(telemetryName, telemetryValue)
+	telemetry.Properties = metric.Tags()
+	a.addContextTags(metric, telemetry)
+	telemetry.Timestamp = metric.Time()
+	return telemetry
+}
+
+func (a *ApplicationInsights) createAggregateMetricTelemetry(metric telegraf.Metric) (*appinsights.AggregateMetricTelemetry, []string) {
+	var telemetryValue float64
+	var telemetryCount int
+	usedFields := make([]string, 0, 6) // We will use up to 6 fields
+	var err error
+
+	// Get the sum of all individual measurements(mandatory property)
+	err = setFloat64TelemetryProperty(&telemetryValue, []string{"sum", "value"}, metric, &usedFields)
+	if err != nil {
+		return nil, nil
+	}
+
+	// Get the count of measurements (mandatory property)
+	err = setIntTelemetryProperty(&telemetryCount, []string{"count", "samples"}, metric, &usedFields)
+	if err != nil {
+		return nil, nil
+	}
+
+	telemetry := appinsights.NewAggregateMetricTelemetry(metric.Name())
+	telemetry.Value = telemetryValue
+	telemetry.Count = telemetryCount
+	telemetry.Properties = metric.Tags()
+	a.addContextTags(metric, telemetry)
+	telemetry.Timestamp = metric.Time()
+
+	// We attempt to set min, max, variance and stddev fields but do not really care if they are not present--
+	// they are not essential for aggregate metric.
+	// By convention AppInsights prefers stddev over variance, so to be consistent, we test for stddev after testing for variance.
+	setFloat64TelemetryProperty(&(telemetry.Min), []string{"min"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Max), []string{"max"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Variance), []string{"variance"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.StdDev), []string{"stddev"}, metric, &usedFields)
+
+	return telemetry, usedFields
+}
+
+func (a *ApplicationInsights) createTelemetryForUnusedFields(metric telegraf.Metric, usedFields []string) []appinsights.Telemetry {
+	fields := metric.Fields()
+	retval := make([]appinsights.Telemetry, 0, len(fields))
+
+	for fieldName := range fields {
+		if contains(usedFields, fieldName) {
+			continue
+		}
+
+		telemetry := a.createSimpleMetricTelemetry(metric, fieldName, true)
+		if telemetry != nil {
+			retval = append(retval, telemetry)
+		}
+	}
+
+	return retval
+}
+
+func (a *ApplicationInsights) addContextTags(metric telegraf.Metric, telemetry appinsights.Telemetry) {
+	for contextTagName, tagSourceName := range a.ContextTagSources {
+		if contextTagValue, found := metric.GetTag(tagSourceName); found {
+			telemetry.ContextTags()[contextTagName] = contextTagValue
+		}
+	}
+}
+
+func setFloat64TelemetryProperty(
+	telemetryProperty *float64,
+	candidateFields []string,
+	metric telegraf.Metric,
+	usedFields *[]string) error {
```

Comment:
I would have this function return `(float64, error)` and return the telemetryPropery.  No need to pass the usedFields by pointer it will work the same.

Reply:
I am not sure if applying this suggestion is an improvement

If the `usedFields` slice is not passed as a pointer, the modifications made to it will not be visible outside of `setFloat64TelemetryProperty` (which is required)

As to returning `(float64, error)` this will complicate consumer code. The point of the method is that the modification succeeds _sometimes, but not always_. If I return a value and error tuple, I will need an extra variable to grab the returned value, check the error, and if the error is nil, set the telemetry property to the actual value. This is more code than just calling `setFloat64TelemetryProperty` passing the address of the telemetry property to set. 

There is a few places in the code where the property set is a "best effort" and `setFloat64TelemetryProperty` is especially useful in this case because one can just call it and ignore the returned error.

Reply:
Slices work like a struct containing a pointer to the underlying data, so no copy will be made and the changes will be visible.  It will work either way but passing a pointer to a slice is not Go convention, and is never used in the Go standard library.

On returning `(float64, error)`, again it is just Go convention.  Keep in mind that float64 types are already set to a zero value and so you would only have to conditionally assign if you have set a default value that you want to preserve.

Reply:
Maybe I am misunderstanding what you are saying regarding the slice point, but run this on Golang play--there is a difference:

```
package main

import (
	"fmt"
)

func main() {
  var sl []string
  changeSlice(sl, "Foo")
  fmt.Printf("%v\n", sl)
  changeSlicePointer(&sl, "Bar")
  fmt.Printf("%v\n", sl)
}

func changeSlice(sl []string, new string) {
  sl = append(sl, new)
}

func changeSlicePointer(sl *[]string, new string) {
  *sl = append(*sl, new)
}
```

Good point regarding zero value, I will change that

Reply:
I see, because we are appending.  I guess the Go way would probably be to return the new slice (just as append does) but I'll let you decide.

Reply:
OK, I have changed the return value but kept the slice update via pointer argument